docs/release-notes/mongoid-8.0.txt

```diff
@@ -47,6 +47,40 @@ changed in Mongoid 8.0:
 Please refer to :ref:`configuration option <configuration-options>` for
 the description and effects of each of these options.
 
+Storing Uncastable Types and the ``validate_attribute_types`` Flag
+------------------------------------------------------------------
+
+In Mongoid 8, Mongoid deals with "uncastable values" based on the values of the
+``validates_attribute_types`` flag. If the ``validates_attribute_types`` flag
+is set to true, a ``Mongoid:Errors::InvalidValue`` exception is raised, and if it is set to false
+a ``nil`` is written. See the secion on :ref:`Uncastable Values <uncastable-values>`
+for more details.
+
+Some ``mongoize`` methods were also changed to perform consistently with rails
+and the other mongoize methods. The following is a table of the changes in
+functionality:
+
++--------------+------------------------+------------------------+-------------------+
+| Field Type   | Situation              | Previous Functionality | New Functionality |
++==============+========================+========================+===================+
+| Integer/Float| When a non-numeric     | return ``nil``         | return ``42``     |
+|              | string starts with a   |                        |                   |
+|              | number: "42bogus"      |                        |                   |
```

Comment:
Hmmm I disagree with this. Bogus chars should make the whole thing bogus.

Reply:
Hey Johnny, originally I thought so too, but this is rails behavior since it is the result of `to_i`. We've decided to go with the rails behavior for consistency.

Reply:
When it comes to working with potentially corrupted data, I don't think that "Doing what Rails happens to do" is a sensible barometer. There's no reason to change Mongoid's existing behavior in this case; we should error on the side of caution rather than **further corrupting corrupted data.** (125 --> "12h5" --> "12")

Before making this change specifically, I suggest you take a poll of actual users and ask what they think is reasonable.

Reply:
@dblock @yuki24 care to comment on this?

Reply:
I agree with @johnnyshields. Following what rails does could be a sensible default if (and only if) there is really no concern. It does seem like `validates_attribute_types` should make value checking more strict, so implicit casting may not be as strict as it should.

Reply:
We created https://jira.mongodb.org/browse/MONGOID-5382 to implement a mechanism that would avoid potential data loss due to type/value conversions during attribute value assignment. The precedent (as set by Rails since, as far as I know, the beginning - circa 2004-2005) is to discard, partially or completely, incoming data when it doesn't neatly convert into the column type. Mongoid has this behavior right now and the changes we are making presently don't "further corrupt data". We do need to settle on one behavior though, which is either to discard the entire input value in cases like `42foo` or to follow the core language (`to_i` converts `42foo` to `42`). We decided to follow the behavior of core language.

Please add your feedback about how to implement stricter value conversions to https://jira.mongodb.org/browse/MONGOID-5382.

Reply:
@p-mongo in the spirit of minimizing change, I'd like to ask that we remove the `"42bogus" --> 42` change from this PR and add it only if we agree to do so after discussion of MONGOID-5382. Otherwise it feels like a "drive-by" change in this PR. It is the only change in this PR which coerces type; all other changes `return nil or raise InvalidError`.

Reply:
I think I kind of agree, and I've changed this to not mongoize "42bogus" to 42

Reply:
💯 on not corrupting data